### PR TITLE
[Toolchain] Uninstall package before installation

### PR DESCRIPTION
**Pull request by jyoungyun**
Debian package allows to install only one package in one machine.
This commit removes package if the same package is already installed
through interaction with users.

ONE-vscode-DCO-1.0-Signed-off-by: Jiyoung Yun <jy910.yun@samsung.com>

### CODE REVIEW

File: src/Project/ToolRunner.ts [hunk 1]
```diff
@@ -161,7 +161,6 @@ export class ToolRunner {
         // TODO(jyoung): Need password encryption
         tool = `echo ${process.env.userp} | sudo -S ` + tool;
         this.child = cp.spawn(tool, toolargs, {cwd: path, shell: true});
-        process.env.userp = '';
```
Comment (jyoungyun) [inline src/Project/ToolRunner.ts:164]: If there is already installed package, it should be removed first.
In this case, this PR requests the two requests using one job to be entered the password from the user only once. However, this code resets the user password before all requests are complete. So I remove this line and reset the password after all requests are done. (JobRunner.ts L102)

**Comment by jyoungyun** [comment]
* Already installed toolchain
   - Show information message `Backend toolchain can be installed only one. Do you want to remove the existing installed toolchain?`
   - If yes, remove the existing installed toolchain and then install the selected toolchain
   - install_uninstall_yes
   - If no, show canceled information message `Installation is canceled.`
   - install_uninstall_no

from https://github.com/Samsung/ONE-vscode/issues/569#issuecomment-1150720962

File: src/View/InstallQuickInput.ts [hunk 2]
```diff
@@ -157,21 +160,51 @@ export async function showInstallQuickInput() {
...
+        resolve(toolchain);
+      };
+      const failed: JobCallback = function() {
+        reject();
+      };
+
+      const installed =
+          toolchainEnv.listInstalled().filter(value => value instanceof DebianToolchain);
+      if (installed.length > 1) {
+        throw Error('One or more debian toolchains cannot be installed');
```
Comment (dayo09) [inline src/View/InstallQuickInput.ts:176]: How about "Installed debian toolchain must be unique"?

File: src/Project/JobRunner.ts [hunk 3]
```diff
@@ -98,6 +98,9 @@ export class JobRunner extends EventEmitter {
 
   private onCleanup() {
     this.running = false;
+    if (process.env.userp !== '') {
```
Comment (YongseopKim) [inline src/Project/JobRunner.ts:101]: (Optional) This looks okay when `if condition` is removed even.

File: src/View/InstallQuickInput.ts [hunk 4]
```diff
@@ -157,21 +160,51 @@ export async function showInstallQuickInput() {
...
+      const success: JobCallback = function() {
+        resolve(toolchain);
+      };
+      const failed: JobCallback = function() {
+        reject();
+      };
+
+      const installed =
+          toolchainEnv.listInstalled().filter(value => value instanceof DebianToolchain);
+      if (installed.length > 1) {
```
Comment (YongseopKim) [inline src/View/InstallQuickInput.ts:175]: This looks a little complicated. How about this?

```ts
if (installed.length > 1) {
    throw Error('One or more debian toolchains cannot be installed');
}

if (installed.length === 0) {
    toolchainEnv.install(toolchain, success, failed);
    return;
}

const answer = await vscode.window.showInformationMessage(
    'Backend toolchain can be installed only one. Do you want to remove the existing installed toolchain?',
    'Yes', 'No');
if (answer === 'Yes') {
    const jobs: Array<Job> = [];
    const uninstallJob = new JobUninstall(installed[0].uninstall());
    uninstallJob.failureCallback = failed;
    jobs.push(uninstallJob);
    const installJob = new JobInstall(toolchain.install());
    installJob.successCallback = success;
    installJob.failureCallback = failed;
    jobs.push(installJob);
    toolchainEnv.request(jobs);
} else {
    Balloon.info('Installation is canceled.');
    reject();
}
```

File: src/Project/JobRunner.ts (hunk 3 above)
Comment (jyoungyun) [reply src/Project/JobRunner.ts:101]: Yes, right. I just want to express it more accurately. :) I will update it.

**Comment by jyoungyun** [comment]
@dayo09 @YongseopKim PTAL

File: src/View/InstallQuickInput.ts (hunk 4 above)
Comment (YongseopKim) [reply src/View/InstallQuickInput.ts]: FYI

**Review by YongseopKim — APPROVED** [review]
LGTM

**Pull request merged**